Mathlib/Data/Seq/Computation.lean

```diff
@@ -0,0 +1,1261 @@
+    b ∈ bind s f :=
+  let ⟨_, h1⟩ := exists_results_of_mem h1
+  let ⟨_, h2⟩ := exists_results_of_mem h2
+  (results_bind h1 h2).mem
+#align computation.mem_bind Computation.mem_bind
+
+instance terminates_bind (s : Computation α) (f : α → Computation β) [Terminates s]
+    [Terminates (f (get s))] : Terminates (bind s f) :=
+  terminates_of_mem (mem_bind (get_mem s) (get_mem (f (get s))))
+#align computation.terminates_bind Computation.terminates_bind
+
+@[simp]
+theorem get_bind (s : Computation α) (f : α → Computation β) [Terminates s]
+    [Terminates (f (get s))] : get (bind s f) = get (f (get s)) :=
+  get_eq_of_mem _ (mem_bind (get_mem s) (get_mem (f (get s))))
+#align computation.get_bind Computation.get_bind
+
+@[simp]
+theorem length_bind (s : Computation α) (f : α → Computation β) [_T1 : Terminates s]
+    [_T2 : Terminates (f (get s))] : length (bind s f) = length (f (get s)) + length s :=
+  (results_of_terminates _).len_unique <|
+    results_bind (results_of_terminates _) (results_of_terminates _)
+#align computation.length_bind Computation.length_bind
+
+theorem of_results_bind {s : Computation α} {f : α → Computation β} {b k} :
+    Results (bind s f) b k → ∃ a m n, Results s a m ∧ Results (f a) b n ∧ k = n + m := by
+  induction' k with n IH generalizing s <;> apply recOn s (fun a => _) fun s' => _ <;> intro e h
+  · simp [thinkN] at h
+    refine' ⟨e, _, _, results_ret _, h, rfl⟩
+  · have := congr_arg head (eq_thinkN h)
+    contradiction
+  · simp at h
+    refine' ⟨e, _, n + 1, results_ret _, h, rfl⟩
+  · simp at h
+    exact by
+      let ⟨a, m, n', h1, h2, e'⟩ := IH h
+      rw [e'] ; exact ⟨a, m.succ, n', results_think h1, h2, rfl⟩
+#align computation.of_results_bind Computation.of_results_bind
+
+theorem exists_of_mem_bind {s : Computation α} {f : α → Computation β} {b} (h : b ∈ bind s f) :
+    ∃ a ∈ s, b ∈ f a :=
+  let ⟨_, h⟩ := exists_results_of_mem h
+  let ⟨a, _, _, h1, h2, _⟩ := of_results_bind h
+  ⟨a, h1.mem, h2.mem⟩
+#align computation.exists_of_mem_bind Computation.exists_of_mem_bind
+
+theorem bind_promises {s : Computation α} {f : α → Computation β} {a b} (h1 : s ~> a)
+    (h2 : f a ~> b) : bind s f ~> b := fun b' bB => by
+  rcases exists_of_mem_bind bB with ⟨a', a's, ba'⟩
+  rw [← h1 a's] at ba'; exact h2 ba'
+#align computation.bind_promises Computation.bind_promises
+
+instance monad : Monad Computation where
+  map := @map
+  pure := @return'
+  bind := @bind
+
+instance : LawfulMonad Computation := LawfulMonad.mk'
+  (id_map := @map_id)
+  (bind_pure_comp := @bind_ret)
+  (pure_bind := @ret_bind)
+  (bind_assoc := @bind_assoc)
+
+theorem has_map_eq_map {β} (f : α → β) (c : Computation α) : f <$> c = map f c :=
+  rfl
+#align computation.has_map_eq_map Computation.has_map_eq_map
+
+@[simp]
+theorem return'_def (a) : (return' a : Computation α) = return' a :=
+  rfl
+#align computation.return_def Computation.return'_def
```

Comment:
Is this theorem required? As `#lint` says, `return' a` already has type `Computation α`.

Reply:
Grepping in the `mathlib` codebase, this `def` is the only mention.

Reply:
Looking at the Lean 3 source, that first `return'` is a `_root_.return` not a `computation.return` so this lemma is not a tautology. You can't check that the lemma isn't used just by grepping the code base because the simplifier might use it. It should be re-instated in my opinion.

Reply:
This should probably be `pure` not `return'`. In lean 4 `return` is used for early return, and `pure` for the monad operation, while in lean 3 `pure` and `return` were interchangeable.